source/Nevermore/Advanced/WriteTransaction.cs

```diff
@@ -217,5 +236,77 @@ public async Task CommitAsync(CancellationToken cancellationToken = default)
             await Transaction.CommitAsync(cancellationToken);
             await configuration.Hooks.AfterCommitAsync(this);
         }
+
+        object[] ExecuteDataModification(PreparedCommand command)
+        {
+            if (!command.Mapping.IsRowVersioningEnabled)
+            {
+                ExecuteNonQuery(command);
+                return Array.Empty<object>();
+            }
+
+            var newRowVersions = new List<object>();
+            using (var reader = ExecuteReader(command))
+            {
+                while (reader.Read())
+                {
+                    newRowVersions.Add(reader.GetValue(0));
+                }
+            }
+
+            return newRowVersions.ToArray();
+        }
+
+        object ExecuteSingleDataModification(PreparedCommand command)
+        {
+            var result = ExecuteDataModification(command);
+            return result.SingleOrDefault();
+        }
+
+        async Task<object[]> ExecuteDataModificationAsync(PreparedCommand command, CancellationToken cancellationToken)
+        {
+            if (!command.Mapping.IsRowVersioningEnabled)
+            {
+                await ExecuteNonQueryAsync(command, cancellationToken);
+                return Array.Empty<object>();
+            }
+
+            var newRowVersions = new List<object>();
+            using (var reader = await ExecuteReaderAsync(command, cancellationToken))
```

Comment:
To make RowVersioning efficient I need to be able to read data from an insert or update.  Unfortunately, Chaos Monkey causes its regular trouble and tests fail randomly because INSERTs and UPDATEs are not retried. 

@hnrkndrssn you seemed to add that code, any ideas? 

Sample build: https://build.octopushq.com/buildConfiguration/OctopusDeploy_LIbraries_Nevermore/2282122

Reply:
do you need to use `ExecuteReaderWithRetryAsync` maybe?

Reply:
The thing this is an insert that returns data. I should not be retried.  Are we getting a lot of value from Chaos monkey? Can its job achieve differently? Happy to zoom if that's easier.